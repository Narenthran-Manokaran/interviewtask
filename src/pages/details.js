import React,{ Component } from 'react';
import { View, Text, StyleSheet, Linking } from 'react-native';
import { connect } from 'react-redux';
import { Header, Button, Card } from 'react-native-elements';
import generateImage from './../utils/generateImage';

let styles;

export class Details extends Component {
  constructor(props, context) {
    super(props, context);
  }

  leftComponent = () => {
    return(
      <Button
        icon={{ name: "chevron-left", size: 36, color: "white" }}
        onPress={() => this.props.navigation.goBack()}
      />
    )
  }

  render() {
    const { container, card, cardButton } = styles;
    const { state } = this.props.navigation;
    const { params } = state;

    return (
      <View style={container}>
        <Header
          leftComponent={this.leftComponent}
          centerComponent={{ text: params.title, style: { color: '#fff' } }}
          rightComponent={{ icon: 'search', color: '#fff' }}
        />

        <Card
          title={params.title}
          containerStyle={card}
          image={{ uri: generateImage(params) }}>
          <Text style={{marginBottom: 10}}>{params.abstract}</Text>
          <Button
            backgroundColor='#03A9F4'
            buttonStyle={cardButton}
            title='VIEW NOW'
            onPress={ ()=>{ Linking.openURL(params.url)}}
          />
        </Card>
      </View>
    )
  }  
}

styles = StyleSheet.create({
  card: {
    padding: 0,
    margin: 0,
    flex: 1,
    flexDirection: 'column'
  },
  cardButton: {
    borderRadius: 0,
    marginLeft: 0,
    marginRight: 0,
    marginBottom: 0
  },
  container: {
    // marginTop: 100,
    flex: 1,
    backgroundColor: 'lightgray'
  },
  text: {
    fontSize: 12,
  },
  button: {
    height: 60,
    margin: 10,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#0b7eff'
  },
  buttonText: {
    color: 'white'
  },
  mainContent: {
    margin: 0,
  },
  image: {
    borderRadius: 37.5,
  },
  contentContainer: {
    alignItems: 'center',
    flex: 1,
    flexDirection: 'row',
    paddingHorizontal: 10,
  },
  textContainer: {
    padding: 10,
    flexWrap: 'wrap',
    flex: 1,
  },
  title: {
    fontSize: 12,
    fontWeight: '700',
    marginBottom: 8
  }
})

export const mapStateToProps = (state) => {
  return {
    appData: state.appData
  }
}

export default connect(mapStateToProps)(Details)