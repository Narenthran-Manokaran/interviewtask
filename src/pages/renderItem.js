import React,{ Component } from 'react';
import { TouchableOpacity, View, Text, StyleSheet, ActivityIndicator, FlatList } from 'react-native';
import { Header, Button, Avatar, Icon, Card } from 'react-native-elements';
import generateImage from './../utils/generateImage';

export default renderItem = ({styles, openDetails, item}) => (
    <TouchableOpacity onPress={() => openDetails(item)}>
        <Card containerStyle={styles.card}>
            <View style={styles.contentContainer}>
                <Avatar rounded source={{ uri: generateImage(item) }} size="large" />
                <View style={styles.textContainer}>
                <Text style={styles.title} numberOfLines={2} ellipsizeMode="tail">{item.title}</Text>
                <Text style={styles.text} numberOfLines={1} ellipsizeMode="tail">{item.byline}</Text>
                <View style={styles.cardContent}>
                    <Text style={[styles.text, { flex: 1 }]}>{item.type}</Text>
                    <Icon name='calendar' type='font-awesome' color='#517fa4' size={14} />
                    <Text style={[styles.text, { paddingLeft: 5 } ]}>{new Date(item.published_date).toDateString()}</Text>
                </View>
                </View>
                <View style={{ justifyContent: 'center' }}>
                <Icon name='angle-right' type='font-awesome' color='#517fa4' size={24} />
                </View>
            </View>
        </Card>
    </TouchableOpacity>
)

