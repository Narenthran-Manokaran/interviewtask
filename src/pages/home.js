import React,{ Component } from 'react';
import { View, Text, StyleSheet, ActivityIndicator, FlatList } from 'react-native';
import { connect } from 'react-redux';
import { Header, Button } from 'react-native-elements';
import { DrawerActions, Header as AppHeader } from 'react-navigation';
import { NavigationActions } from 'react-navigation';
import renderItem from './renderItem';
import { NavigationEvents } from "react-navigation";
import { fetchData } from './../actions';

let styles
export class Home extends Component {
  constructor(props, context) {
    super(props, context);
  }

  openDetails = (params) => {
    const navigateAction = NavigationActions.navigate({
      routeName: 'details',
      params,
    });
    this.props.navigation.dispatch(navigateAction);
  }
  keyExtractor = (item, index) => item.title

  leftComponent = () => <Button icon={{ name: "menu", size: 24, color: "white" }} onPress={() => this.props.navigation.dispatch(DrawerActions.openDrawer())} />

  renderItem = ({ item }) => renderItem({styles, openDetails: this.openDetails, item })
  
  render() {
    const { container, mainContent} = styles;
    const { appData, navigation } = this.props;

    return (
      <View style={container}>
        <Header
          leftComponent={this.leftComponent}
          centerComponent={{ text: navigation.state.params.title, style: { color: '#fff' } }}
          rightComponent={{ icon: 'search', color: '#fff' }}
        />
        {appData.error && <Text>Something went wrong.</Text>}
        {appData.isFetching && 
          <View style={[styles.loadingContainer, styles.horizontal]}>
            <ActivityIndicator size="large" color="#0000ff" />
          </View>}
        <View style={mainContent}>
            <FlatList
              data={appData.data}
              renderItem={this.renderItem}
              keyExtractor={this.keyExtractor}
            />
        </View>
        <NavigationEvents onWillFocus={payload => (payload.action.type !== "Navigation/BACK") && this.props.fetchData(payload.state.params.uri) } />
      </View>
    )
  }  
}

styles = StyleSheet.create({
  card: {
    padding: 0,
    margin: 0
  },
  cardContent: {
    flexDirection: 'row',
    flex: 1,
    alignItems: 'center'
  },
  container: {
    flex: 1,
    backgroundColor: 'lightgray',
  },
  loadingContainer: {
    justifyContent: 'center',
    flex: 1
  },
  horizontal: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    padding: 10,
    flex: 1,
  },
  text: {
    fontSize: 12,
    fontFamily: 'System',
    lineHeight: 18
  },
  button: {
    height: 60,
    margin: 10,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#0b7eff'
  },
  buttonText: {
    color: 'white'
  },
  mainContent: {
    margin: 0,
  },
  image: {
    borderRadius: 37.5,
  },
  contentContainer: {
    alignItems: 'center',
    flex: 1,
    flexDirection: 'row',
    paddingHorizontal: 10,
  },
  textContainer: {
    padding: 10,
    flexWrap: 'wrap',
    flex: 1,
  },
  title: {
    fontSize: 12,
    fontWeight: '700',
    marginBottom: 8,
    fontFamily: 'System'
  }
})

export const mapStateToProps = (state) => {
  return {
    appData: state.appData
  }
}

export const mapDispatchToProps = (dispatch) => {
  return {
    fetchData: (uri) => dispatch(fetchData(uri))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Home)