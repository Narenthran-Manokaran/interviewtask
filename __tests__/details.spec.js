import React from 'react';
import { Details, mapStateToProps } from './../src/pages/details';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Header, Button, Card } from 'react-native-elements';

configure({ adapter: new Adapter() });

jest.mock('react-native-gesture-handler', () => {});
jest.mock('react-navigation-stack', () => { return {Header: ()=>'whatever'} });

const initialState = {
  data: [],
  dataFetched: false,
  isFetching: false,
  error: false,
}

const navigation = {
  state: {
    params: {
      title: 'Test'
    }
  }
}

describe('Details', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Details appData={initialState} navigation={navigation}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should show previously rolled value', () => {
    expect(mapStateToProps({appData:initialState})).toEqual({appData:initialState});
  });

  const renderedComponent = shallow(<Details appData={initialState} navigation={navigation}/>);

  it('should contain 1 CardItems', () => {
    const items = renderedComponent.find(Card);
    expect(items).toHaveLength(1);
  });
  it('should contain 1 HeaderItems', () => {
    const items = renderedComponent.find(Header);
    expect(items).toHaveLength(1);
  });

  test('test onPress functionality', () => {
    const onPressEvent = jest.fn();
    onPressEvent.mockReturnValue('Link on press invoked');
    const wrapper = shallow(<Details appData={initialState} navigation={navigation} onPress={ onPressEvent } text='CustomLink Component'/>);
    wrapper.find(Button).first().props().onPress();
    expect(onPressEvent.mock.calls.length).toBe(0);
  });
})