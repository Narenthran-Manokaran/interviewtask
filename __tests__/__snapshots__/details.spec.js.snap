// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "lightgray",
      "flex": 1,
    }
  }
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#2089dc",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "height": 64,
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingTop": 20,
      }
    }
  >
    <Image
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": 64,
            "width": undefined,
          },
          undefined,
        ]
      }
      testID="headerContainer"
    />
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "borderRadius": 3,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#2089dc",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <View
              style={
                Object {
                  "marginHorizontal": 5,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                theme={
                  Object {
                    "colors": Object {
                      "disabled": "hsl(208, 8%, 90%)",
                      "divider": "#bcbbc1",
                      "error": "#ff190c",
                      "grey0": "#393e42",
                      "grey1": "#43484d",
                      "grey2": "#5e6977",
                      "grey3": "#86939e",
                      "grey4": "#bdc6cf",
                      "grey5": "#e1e8ee",
                      "greyOutline": "#bbb",
                      "platform": Object {
                        "android": Object {
                          "error": "#f44336",
                          "primary": "#2196f3",
                          "secondary": "#9C27B0",
                          "success": "#4caf50",
                          "warning": "#ffeb3b",
                        },
                        "ios": Object {
                          "error": "#ff3b30",
                          "primary": "#007aff",
                          "secondary": "#5856d6",
                          "success": "#4cd964",
                          "warning": "#ffcc00",
                        },
                      },
                      "primary": "#2089dc",
                      "searchBg": "#303337",
                      "secondary": "#8F0CE8",
                      "success": "#52c41a",
                      "warning": "#faad14",
                    },
                  }
                }
                underlayColor="white"
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "white",
                        "fontSize": 36,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="iconIcon"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 3,
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          Object {
            "color": "#fff",
          }
        }
        text="Test"
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
      >
        Test
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-end",
          }
        }
      >
        <View
          icon="search"
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            }
          }
          theme={
            Object {
              "colors": Object {
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "primary": "#2196f3",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "primary": "#007aff",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#8F0CE8",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            }
          }
          underlayColor="white"
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontSize": 24,
                },
                Object {
                  "backgroundColor": "transparent",
                },
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="iconIcon"
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderColor": "#e1e8ee",
        "borderWidth": 1,
        "flex": 1,
        "flexDirection": "column",
        "margin": 0,
        "marginBottom": 0,
        "padding": 0,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#43484d",
              "fontSize": 17.5,
              "fontWeight": "bold",
              "marginBottom": 15,
              "marginTop": 15,
              "textAlign": "center",
            }
          }
          testID="cardTitle"
          theme={
            Object {
              "colors": Object {
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "primary": "#2196f3",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "primary": "#007aff",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#8F0CE8",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            }
          }
        >
          Test
        </Text>
      </View>
      <View>
        <View
          accessibilityIgnoresInvertColors={true}
          style={
            Array [
              Object {
                "height": 150,
                "width": null,
              },
              undefined,
            ]
          }
        >
          <Image
            source={
              Object {
                "uri": "https://cmkt-image-prd.global.ssl.fastly.net/0.1.0/ps/1518711/910/607/m1/fpnw/wm0/lawyer-avatar-icon-01-.jpg?1470209857&s=89c9e32338a33862d93a44b76a1ae3e9",
              }
            }
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "height": undefined,
                  "width": undefined,
                },
                undefined,
              ]
            }
          />
        </View>
        <View
          style={
            Object {
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 10,
              }
            }
          />
          <View
            style={
              Object {
                "borderRadius": 3,
              }
            }
          >
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#2089dc",
                    "borderColor": "#2089dc",
                    "borderRadius": 0,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginBottom": 0,
                    "marginLeft": 0,
                    "marginRight": 0,
                    "padding": 8,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 18,
                      "paddingBottom": 1,
                      "paddingTop": 2,
                      "textAlign": "center",
                    }
                  }
                >
                  VIEW NOW
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
